nixos/modules/services/desktop-managers/cosmic.nix

```
@@ -32,45 +67,32 @@ in
       "/share/backgrounds"
       "/share/cosmic"
     ];
-    environment.systemPackages =
-      with pkgs;
-      [
-        adwaita-icon-theme
-        alsa-utils
-        cosmic-applets
-        cosmic-applibrary
-        cosmic-bg
-        cosmic-comp
-        cosmic-edit
-        cosmic-files
-        config.services.displayManager.cosmic-greeter.package
-        cosmic-icons
-        cosmic-idle
-        cosmic-launcher
-        cosmic-notifications
-        cosmic-osd
-        cosmic-panel
-        cosmic-player
-        cosmic-randr
-        cosmic-screenshot
-        cosmic-session
-        cosmic-settings
-        cosmic-settings-daemon
-        cosmic-term
-        cosmic-wallpapers
-        cosmic-workspaces-epoch
-        hicolor-icon-theme
-        playerctl
-        pop-icon-theme
-        pop-launcher
-        xdg-user-dirs
-      ]
-      ++ lib.optionals cfg.xwayland.enable [
-        xwayland
-      ]
-      ++ lib.optionals config.services.flatpak.enable [
-        cosmic-store
-      ];
+    environment.systemPackages = utils.removePackagesByName (
+      corePkgs
+      ++ (
+        with pkgs;
+        [
+          adwaita-icon-theme
+          alsa-utils
+          cosmic-edit
+          cosmic-icons
+          cosmic-player
+          cosmic-randr
+          cosmic-screenshot
+          cosmic-term
+          cosmic-wallpapers
+          hicolor-icon-theme
+          playerctl
+          pop-icon-theme
+          pop-launcher
+          xdg-user-dirs
+        ]
+        ++ lib.optionals config.services.flatpak.enable [
+          # User may have Flatpaks enabled but might not want the `cosmic-store` package.
+          cosmic-store
+        ]
+      )
+    ) config.environment.cosmic.excludePackages;
```

Instead of adding a warning in case a required package gets removed, we should just not allow them to be removed. Makes the implementation simpler and cleaner, just like it is done in other desktop environment modules such as [Plasma 6](https://github.com/NixOS/nixpkgs/blob/a776364d6758396a3704071cc9f5625ad997eed3/nixos/modules/services/desktop-managers/plasma6.nix#L182-L184)

```suggestion
    environment.systemPackages =
    let
      # corePackages or requiredPackages moved in here
      optionalPackages = with pkgs; [
          adwaita-icon-theme
          alsa-utils
          cosmic-applets
          cosmic-edit
          cosmic-icons
          cosmic-player
          cosmic-randr
          cosmic-settings
          cosmic-screenshot
          cosmic-term
          cosmic-wallpapers
          hicolor-icon-theme
          playerctl
          pop-icon-theme
          pop-launcher
          xdg-user-dirs
      ]
      ++ lib.optional config.services.flatpak.enable cosmic-store;
    in
    corePackages
    ++ utils.removePackagesByName optionalPackages config.environment.cosmic.excludePackages
    ++ lib.optional cfg.xwayland.enable xwayland;
```

---

We still need a warning in this case for "why the user requested a package to be excluded but was not excluded." And we end up issuing a warning regardless. In the current state of the patches, at least the user can remove the packages, at their responsibility.

---

In my honest opinion, I don't think we should let a user remove packages that make the session crash, for example the compositor, the panel, the notifications daemon and the settings daemon. The rest of the packages can be removed without causing a crash, but will result in reduced functionality

---

If we have a general list that will start a session and keep it stable I think that is good. It would give folks moving from Sway or Hyprland a place to start if they want to continue using 3rd party Wayland tools.

---

It is my opinion that we should issue a warning instead of outright error-ing out. Mainly because we do not want to discourage users from substituting their modular packages in place with the first-party packages. I have no idea what someone would substitute a package like `cosmic-panel` with, but that's an option given COSMIC's modularity.

---

I think you're currently thinking a use case exists where one currently does not. If/when a valid substitute for `cosmic-panel` exists, we can always make this not error.

---

Can we have a vote on this? :+1: for keeping it as warning and :-1: as turning it into an error.

---

Current ratio is 4 in favour (1 non-maintainer among them) of a warning and 1 in favour of an error. I'll keep it as a warning.

---

A warning is fine, but in that case there needs to be an option to disable the warning. We should never get into the situation where users are expecting there to always be a warning when they evaluate their configuration — there needs to be a way to fix the warning.  Otherwise we cause alarm fatigue.

---

Addressed in the latest force push.

---

Should probably actually mention that option in the warning messaeg.